### PR TITLE
Support using __str__ for version not just strings

### PR DESCRIPTION
**Pull request by sontek**

### CODE REVIEW

**Comment by msabramo** [comment]
:+1: after the test failures are fixed.

**Comment by sontek** [comment]
@msabramo Tests fixed :)

**Comment by msabramo** [comment]
:+1: 

Some context/motivation: This would fix the issue at https://bitbucket.org/pypa/setuptools/issues/409/unable-to-build-wheel-from-unittest2

**Comment by sontek** [comment]
Yeah, in general this type of version should have been picked up by `LegacyVersion` and then it would have "just worked" :tm: but since `setuptools` uses `Version` directly instead of `parse` everything blows up.

I think the solution is probably a mix of both... We should allow `Version` to use `__str__` but we should also update `setuptools` to call `parse` instead.

File: tests/test_version.py [hunk 1]
```diff
@@ -33,10 +33,22 @@ def test_parse(version, klass):
...
+class VersionObj(object):
+    def __init__(self, version):
+        self.version = version
+
+    def __str__(self):
+        return self.version
+
+    def __repr__(self):
+        return repr(self.version)
+
```
Comment (rbtcollins) [inline tests/test_version.py]: nit: PEP8 - 2 lines of VWS between top level definitions

**Comment by rbtcollins** [comment]
So a few things I guess. I'm a maintainer of unittest2, and I've had no problem building wheels. Not sure why that is.
Secondly, I'm probably going to move unittest2 off of hg.python.org to github and switch it to using pbr, at which point all this lazy evil will just go away completely. So perhaps worth just waiting for that.

lastly, I hate the evil that I did, I really don't encourage supporting it. I build and distribute wheels just fine, so there's also something confounding things here.

File: tests/test_version.py (hunk 1 above)
Comment (sontek) [inline tests/test_version.py]: The tox is running pep8 and it didn't complain:

https://travis-ci.org/pypa/packaging/jobs/70176536

**Comment by sontek** [comment]
@rbtcollins Upgrade your pip/setuptools/wheel and you'll see the problem.  I didn't check which specific version doesn't work with it but the latest of everything definitely doesn't.

**Comment by rbtcollins** [comment]
I am using latest everything:

```
unittest2.hg$ python setup.py sdist bdist_wheel 
/home/robertc/.virtualenvs/unittest2-2.6-subunit-lucid/lib/python2.6/site-packages/setuptools/dist.py:294: UserWarning: The version specified (<__main__.late_version instance at 0xf74196cc>) is an invalid version, this may not work as expected with newer versions of setuptools, pip, and PyPI. Please see PEP 440 for more details.
  "details." % self.metadata.version
running sdist
running egg_info
writing requirements to unittest2.egg-info/requires.txt
writing unittest2.egg-info/PKG-INFO
writing top-level names to unittest2.egg-info/top_level.txt
writing dependency_links to unittest2.egg-info/dependency_links.txt
writing entry points to unittest2.egg-info/entry_points.txt
reading manifest file 'unittest2.egg-info/SOURCES.txt'
writing manifest file 'unittest2.egg-info/SOURCES.txt'
creating unittest2-1.2.0
creating unittest2-1.2.0/unittest2
creating unittest2-1.2.0/unittest2.egg-info
creating unittest2-1.2.0/unittest2/test
making hard links in unittest2-1.2.0...
hard linking README.txt -> unittest2-1.2.0
hard linking setup.cfg -> unittest2-1.2.0
hard linking setup.py -> unittest2-1.2.0
hard linking unittest2/__init__.py -> unittest2-1.2.0/unittest2
hard linking unittest2/__main__.py -> unittest2-1.2.0/unittest2
hard linking unittest2/case.py -> unittest2-1.2.0/unittest2
hard linking unittest2/collector.py -> unittest2-1.2.0/unittest2
hard linking unittest2/compatibility.py -> unittest2-1.2.0/unittest2
hard linking unittest2/loader.py -> unittest2-1.2.0/unittest2
hard linking unittest2/main.py -> unittest2-1.2.0/unittest2
hard linking unittest2/result.py -> unittest2-1.2.0/unittest2
hard linking unittest2/runner.py -> unittest2-1.2.0/unittest2
hard linking unittest2/signals.py -> unittest2-1.2.0/unittest2
hard linking unittest2/suite.py -> unittest2-1.2.0/unittest2
hard linking unittest2/util.py -> unittest2-1.2.0/unittest2
hard linking unittest2.egg-info/PKG-INFO -> unittest2-1.2.0/unittest2.egg-info
hard linking unittest2.egg-info/SOURCES.txt -> unittest2-1.2.0/unittest2.egg-info
hard linking unittest2.egg-info/dependency_links.txt -> unittest2-1.2.0/unittest2.egg-info
hard linking unittest2.egg-info/entry_points.txt -> unittest2-1.2.0/unittest2.egg-info
hard linking unittest2.egg-info/requires.txt -> unittest2-1.2.0/unittest2.egg-info
hard linking unittest2.egg-info/top_level.txt -> unittest2-1.2.0/unittest2.egg-info
hard linking unittest2/test/__init__.py -> unittest2-1.2.0/unittest2/test
hard linking unittest2/test/_test_unittest2_with.py -> unittest2-1.2.0/unittest2/test
hard linking unittest2/test/dummy.py -> unittest2-1.2.0/unittest2/test
hard linking unittest2/test/support.py -> unittest2-1.2.0/unittest2/test
hard linking unittest2/test/test_assertions.py -> unittest2-1.2.0/unittest2/test
hard linking unittest2/test/test_break.py -> unittest2-1.2.0/unittest2/test
hard linking unittest2/test/test_case.py -> unittest2-1.2.0/unittest2/test
hard linking unittest2/test/test_discovery.py -> unittest2-1.2.0/unittest2/test
hard linking unittest2/test/test_functiontestcase.py -> unittest2-1.2.0/unittest2/test
hard linking unittest2/test/test_loader.py -> unittest2-1.2.0/unittest2/test
hard linking unittest2/test/test_new_tests.py -> unittest2-1.2.0/unittest2/test
hard linking unittest2/test/test_program.py -> unittest2-1.2.0/unittest2/test
hard linking unittest2/test/test_result.py -> unittest2-1.2.0/unittest2/test
hard linking unittest2/test/test_runner.py -> unittest2-1.2.0/unittest2/test
hard linking unittest2/test/test_setups.py -> unittest2-1.2.0/unittest2/test
hard linking unittest2/test/test_skipping.py -> unittest2-1.2.0/unittest2/test
hard linking unittest2/test/test_suite.py -> unittest2-1.2.0/unittest2/test
hard linking unittest2/test/test_unittest2_with.py -> unittest2-1.2.0/unittest2/test
copying setup.cfg -> unittest2-1.2.0
Writing unittest2-1.2.0/setup.cfg
tar -cf dist/unittest2-1.2.0.tar unittest2-1.2.0
gzip -f9 dist/unittest2-1.2.0.tar
removing 'unittest2-1.2.0' (and everything under it)
running bdist_wheel
running build
running build_py
installing to build/bdist.linux-i686/wheel
running install
running install_lib
creating build/bdist.linux-i686/wheel
creating build/bdist.linux-i686/wheel/unittest2
creating build/bdist.linux-i686/wheel/unittest2/test
copying build/lib.linux-i686-2.6/unittest2/test/test_suite.py -> build/bdist.linux-i686/wheel/unittest2/test
copying build/lib.linux-i686-2.6/unittest2/test/_test_unittest2_with.py -> build/bdist.linux-i686/wheel/unittest2/test
copying build/lib.linux-i686-2.6/unittest2/test/dummy.py -> build/bdist.linux-i686/wheel/unittest2/test
copying build/lib.linux-i686-2.6/unittest2/test/test_skipping.py -> build/bdist.linux-i686/wheel/unittest2/test
copying build/lib.linux-i686-2.6/unittest2/test/__init__.py -> build/bdist.linux-i686/wheel/unittest2/test
copying build/lib.linux-i686-2.6/unittest2/test/test_break.py -> build/bdist.linux-i686/wheel/unittest2/test
copying build/lib.linux-i686-2.6/unittest2/test/test_runner.py -> build/bdist.linux-i686/wheel/unittest2/test
copying build/lib.linux-i686-2.6/unittest2/test/test_case.py -> build/bdist.linux-i686/wheel/unittest2/test
copying build/lib.linux-i686-2.6/unittest2/test/test_new_tests.py -> build/bdist.linux-i686/wheel/unittest2/test
copying build/lib.linux-i686-2.6/unittest2/test/test_loader.py -> build/bdist.linux-i686/wheel/unittest2/test
copying build/lib.linux-i686-2.6/unittest2/test/test_unittest2_with.py -> build/bdist.linux-i686/wheel/unittest2/test
copying build/lib.linux-i686-2.6/unittest2/test/test_assertions.py -> build/bdist.linux-i686/wheel/unittest2/test
copying build/lib.linux-i686-2.6/unittest2/test/test_result.py -> build/bdist.linux-i686/wheel/unittest2/test
copying build/lib.linux-i686-2.6/unittest2/test/test_functiontestcase.py -> build/bdist.linux-i686/wheel/unittest2/test
copying build/lib.linux-i686-2.6/unittest2/test/test_discovery.py -> build/bdist.linux-i686/wheel/unittest2/test
copying build/lib.linux-i686-2.6/unittest2/test/test_program.py -> build/bdist.linux-i686/wheel/unittest2/test
copying build/lib.linux-i686-2.6/unittest2/test/test_setups.py -> build/bdist.linux-i686/wheel/unittest2/test
copying build/lib.linux-i686-2.6/unittest2/test/support.py -> build/bdist.linux-i686/wheel/unittest2/test
copying build/lib.linux-i686-2.6/unittest2/signals.py -> build/bdist.linux-i686/wheel/unittest2
copying build/lib.linux-i686-2.6/unittest2/result.py -> build/bdist.linux-i686/wheel/unittest2
copying build/lib.linux-i686-2.6/unittest2/main.py -> build/bdist.linux-i686/wheel/unittest2
copying build/lib.linux-i686-2.6/unittest2/__init__.py -> build/bdist.linux-i686/wheel/unittest2
copying build/lib.linux-i686-2.6/unittest2/loader.py -> build/bdist.linux-i686/wheel/unittest2
copying build/lib.linux-i686-2.6/unittest2/runner.py -> build/bdist.linux-i686/wheel/unittest2
copying build/lib.linux-i686-2.6/unittest2/compatibility.py -> build/bdist.linux-i686/wheel/unittest2
copying build/lib.linux-i686-2.6/unittest2/__main__.py -> build/bdist.linux-i686/wheel/unittest2
copying build/lib.linux-i686-2.6/unittest2/suite.py -> build/bdist.linux-i686/wheel/unittest2
copying build/lib.linux-i686-2.6/unittest2/collector.py -> build/bdist.linux-i686/wheel/unittest2
copying build/lib.linux-i686-2.6/unittest2/util.py -> build/bdist.linux-i686/wheel/unittest2
copying build/lib.linux-i686-2.6/unittest2/case.py -> build/bdist.linux-i686/wheel/unittest2
running install_egg_info
Copying unittest2.egg-info to build/bdist.linux-i686/wheel/unittest2-1.2.0-py2.6.egg-info
running install_scripts
creating build/bdist.linux-i686/wheel/unittest2-1.2.0.dist-info/WHEEL
```
and
```
$ pip list
argparse (1.3.0)
linecache2 (1.0.0)
pip (7.1.0)
setuptools (18.0.1)
six (1.9.0)
traceback2 (1.0.0)
unittest2 (1.2.0)
wheel (0.24.0)
```

**Comment by sontek** [comment]
@rbtcollins Thats a different use case, try just running:

```
pip wheel unittest2 --no-use-wheel
```

**Comment by msabramo** [comment]
@rbtcollins :+1: to moving unittest2 to GitHub.

File: tests/test_version.py (hunk 1 above)
Comment (sigmavirus24) [inline tests/test_version.py]: FWIW, this may be a bug in pep8 the tool. Guido has filed a couple bugs around this.

**Comment by domenkozar** [comment]
I've also hit this one using `unittest2`, could it be merged?

**Comment by dstufft** [comment]
I'm going to close this PR, I think that ``packaging`` should not support anything but things that are *actually* strings (and due to the nature of Python, everything has a ``__str__`` method so ``str()`` will never fail).

**Comment by domenkozar** [comment]
@dstufft should we instead validate `version` and fail if it's not a string?

**Comment by dstufft** [comment]
A type check would be fine by me.

**Pull request closed without merging**